Interop: Compiler and C++ Bridging header disagree on ABI of `Optional<CxxValueType>`

### Description

When writing a swift function that returns a small optional C++ value type, the swift compiler seems to generates code to
- assume that the calling scope's return slot has space for the type + a bool (?)
- construct the C++ type by calling its constructor
- return the `Optional<CxxValueType>` in a register.

On the other hand, the generated bridging header from -emit-clang-header-path seems to think the type will be returned as a heap allocated allocated objective-c type (?) with a type metadata and everything that goes along with that.

### Reproduction

The reproduction for this requires a few files:
- a C++ type
- a swift function to return `CxxType?` from that function
- a C++ function that calls the swift function

I've created a project here that can reproduce the issue https://github.com/ADKaster/swift-shlib-cxx-exe with 
```
cmake -B build -GNinja && cmake --build build
```

The most important files are below:

Color.h
```c++
#pragma once

#include <stdint.h>

class Color
{
public:
    Color() = default;
    explicit Color(uint32_t rep) : m_rep(rep) { }

    static bool fromString(Color&, char const*);

private: 
    uint32_t m_rep { 0 };
};
```

Color.cpp
```c++
#include "Color.h"

#include <swift/bridging>
#include <shlib-Swift.h>

bool Color::fromString(Color& color, char const* str)
{
    if (auto c = SwiftShlib::colorFromString(str)) {
        color = c.get();
        return true;
    }
    return false;
}

```

Color.swift
```swift
import shlib

public func colorFromString(_ str: String) -> shlib.Color?
{
    if (str == "red") {
        return Color(1234)
    }
    return nil
}
```

Building with `-DCMAKE_BUILD_TYPE=Debug`, this produces the following linker error:

```
[1/2] Linking Swift shared library shlib/libshlib.so
FAILED: shlib/libshlib.so 
: && /home/andrew/swift/build/buildbot_linux/none-swift_package_sandbox_linux-x86_64/usr/bin/swiftc -j 32 -num-threads 32 -emit-library -Onone -g  -Xlinker --no-allow-shlib-undefined -Xlinker --no-undefined /home/andrew/swift/build/buildbot_linux/none-swift_package_sandbox_linux-x86_64/usr/lib/swift/linux/x86_64/swiftrt.o -use-ld=lld  -Xlinker -soname -Xlinker libshlib.so -o shlib/libshlib.so shlib/CMakeFiles/shlib.dir/Color.cpp.o shlib/CMakeFiles/shlib.dir/Color.swift.o -L /home/andrew/swift/build/buildbot_linux/none-swift_package_sandbox_linux-x86_64/usr/lib/swift/linux   -L /home/andrew/swift/build/buildbot_linux/none-swift_package_sandbox_linux-x86_64/usr/lib/swift/linux/x86_64   -L /usr/lib/gcc/x86_64-linux-gnu/12 -lstdc++  -lm  -lgcc_s  -lgcc  -lc  -lgcc_s  -lgcc && :
error: link command failed with exit code 1 (use -v to see invocation)
ld.lld: error: undefined symbol: $sSo5ColorVMa
>>> referenced by shlib-Swift.h:5765 (shlib/include/shlib-Swift.h:5765)
>>>               /home/andrew/ladybird-org/testing/swift-tests/swift-shlib-cxx-exe/build/shlib/CMakeFiles/shlib.dir/Color.cpp.o:(swift::TypeMetadataTrait<Color>::getTypeMetadata())
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

with `-DCMAKE_BUILD_TYPE=Release`, the error is a bit different and less helpful:

```
error: link command failed with exit code 1 (use -v to see invocation)
ld.lld: error: undefined symbol: $sSo5ColorVMa
>>> referenced by Color.cpp
>>>               /home/andrew/ladybird-org/testing/swift-tests/swift-shlib-cxx-exe/build/shlib/CMakeFiles/shlib.dir/Color.cpp.o:(Color::fromString(Color&, char const*))
>>> referenced by Color.cpp
>>>               /home/andrew/ladybird-org/testing/swift-tests/swift-shlib-cxx-exe/build/shlib/CMakeFiles/shlib.dir/Color.cpp.o:(Color::fromString(Color&, char const*))
>>> referenced by Color.cpp
>>>               /home/andrew/ladybird-org/testing/swift-tests/swift-shlib-cxx-exe/build/shlib/CMakeFiles/shlib.dir/Color.cpp.o:(Color::fromString(Color&, char const*))
>>> referenced 1 more times
```

### Expected behavior

The program should link successfully. getTypeMetadata should not be called when the swift compiler does not actually emit type metadata.

### Environment

$ /home/andrew/swift/build/buildbot_linux/none-swift_package_sandbox_linux-x86_64/usr/bin/swiftc  -version
Swift version 6.0-dev (LLVM 3427f33b53c666b, Swift 225809e3f8e9c7b)
Target: x86_64-unknown-linux-gnu

Also tested with:

$ swiftc -version
swift-driver version: 1.111.2 Apple Swift version 6.0 (swiftlang-6.0.0.5.15 clang-1600.0.22.6)
Target: arm64-apple-macosx14.0

### Additional information

https://forums.swift.org/t/cxxinterop-swift-function-returning-optional-c-type-in-shlib-doesnt-generate-metadata/73423

--- a/lib/IRGen/IRGenSIL.cpp
+++ b/lib/IRGen/IRGenSIL.cpp
@@ -18,6 +18,7 @@
 #include "GenKeyPath.h"
 #include "swift/AST/ASTContext.h"
 #include "swift/AST/ASTMangler.h"
+#include "swift/AST/Decl.h"
 #include "swift/AST/DiagnosticsIRGen.h"
 #include "swift/AST/ExtInfo.h"
 #include "swift/AST/GenericEnvironment.h"
@@ -26,6 +27,9 @@
 #include "swift/AST/Pattern.h"
 #include "swift/AST/SemanticAttrs.h"
 #include "swift/AST/SubstitutionMap.h"
+#include "swift/AST/Type.h"
+#include "swift/AST/TypeExpansionContext.h"
+#include "swift/AST/TypeVisitor.h"
 #include "swift/AST/Types.h"
 #include "swift/Basic/Assertions.h"
 #include "swift/Basic/ExternalUnion.h"
@@ -2472,6 +2476,53 @@ static void emitDynamicSelfMetadata(IRGenSILFunction &IGF) {
   IGF.setDynamicSelfMetadata(selfTy, isExact, value, selfKind);
 }
 
+/// C++ interop might refer to the type metadata in some scenarios.
+/// This function covers those cases and makes sure metadata is emitted
+/// for the foreign types.
+static void noteUseOfMetadataByCXXInterop(IRGenerator &IRGen,
+                                          const SILFunction *f,
+                                          const TypeExpansionContext &context) {
+  auto type = f->getLoweredFunctionType();
+
+  // Notes the use of foreign types in generic arguments for C++ interop.
+  auto processType = [&](CanType type) {
+    struct Walker : TypeWalker {
+      Walker(IRGenerator &IRGen) : IRGen(IRGen) {}
+
+      Action walkToTypePre(Type ty) override {
+        if (auto *BGT = ty->getAs<BoundGenericType>())
+          genericDepth++;
+        else if (auto *nominal = ty->getAs<NominalType>())
+          noteUseOfTypeMetadata(nominal->getDecl());
+        return Action::Continue;
+      }
+
+      Action walkToTypePost(Type ty) override {
+        if (auto *BGT = ty->getAs<BoundGenericType>())
+          genericDepth--;
+
+        return Action::Continue;
+      }
+
+      void noteUseOfTypeMetadata(NominalTypeDecl *type) {
+        if (genericDepth == 0)
+          return;
+        if (!IRGen.hasLazyMetadata(type) || !type->hasClangNode())
+          return;
+        IRGen.noteUseOfTypeMetadata(type);
+      }
+      IRGenerator &IRGen;
+      int genericDepth = 0;
+    } walker{IRGen};
+    type.walk(walker);
+  };
+
+  for (const auto &param : type->getParameters())
+    processType(param.getArgumentType(IRGen.SIL, type, context));
+  for (const auto &result : type->getResultsWithError())
+    processType(result.getReturnValueType(IRGen.SIL, type, context));
+}
+
 /// Emit the definition for the given SIL constant.
 void IRGenModule::emitSILFunction(SILFunction *f) {
   if (f->isExternalDeclaration())
@@ -2486,6 +2537,12 @@ void IRGenModule::emitSILFunction(SILFunction *f) {
                                    f->getASTContext().SILOpts.EnableSerializePackage) &&
       f->isAvailableExternally())
     return;
+
+  // Type metadata for foreign references is not yet supported on Windows. Bug #76168.
+  if (Context.LangOpts.EnableCXXInterop &&
+      f->getLinkage() == SILLinkage::Public &&
+      !Context.LangOpts.Target.isOSWindows())
+    noteUseOfMetadataByCXXInterop(IRGen, f, TypeExpansionContext(*f));
 
   PrettyStackTraceSILFunction stackTrace("emitting IR", f);
   IRGenSILFunction(*this, f).emitSILFunction();

--- a/test/Interop/CxxToSwiftToCxx/link-cxx-type-metadata-accessor.swift
+++ b/test/Interop/CxxToSwiftToCxx/link-cxx-type-metadata-accessor.swift
@@ -17,6 +17,14 @@ struct CxxStruct {
     int x;
 };
 
+struct CxxStruct2 {
+    inline CxxStruct2(int x) : x(x) {}
+    inline CxxStruct2(const CxxStruct &other) : x(other.x) {}
+    inline ~CxxStruct2() {}
+
+    int x;
+};
+
 //--- module.modulemap
 module CxxTest {
     header "header.h"
@@ -30,6 +38,12 @@ public func retCxxStruct() -> CxxStruct {
     return CxxStruct(2)
 }
 
+#if !os(Windows)
+public func retCxxStruct2() -> CxxStruct2? {
+    return CxxStruct2(2)
+}
+#endif
+
 //--- use-swift-cxx-types.cpp
 
 #include "header.h"
@@ -38,5 +52,8 @@ public func retCxxStruct() -> CxxStruct {
 
 int main() {
   auto x = UseCxx::retCxxStruct();
+#ifndef _WIN32
+  auto y = UseCxx::retCxxStruct2();
+#endif
   return 0;
 }

